### PR TITLE
WIP: Add fns to create Endpoints/Clusters CRDs

### PR DESCRIPTION
**Pull request by dfarrell07**
Signed-off-by: Daniel Farrell <dfarrell@redhat.com>

### CODE REVIEW

File: pkg/subctl/cmd/createbroker.go [hunk 1]
```diff
@@ -80,5 +84,35 @@ var createBrokerCmd = &cobra.Command{
 			panic(err.Error())
 		}
 		fmt.Printf("There are %d pods in the cluster\n", len(pods.Items))
+
+		// Create CRDs
```
Comment (mangelajo) [inline pkg/subctl/cmd/createbroker.go]: I would isolate all the CRD creation/verification to it's own pkg, like pkg/broker or something like that, in a way that later on it's easy to move to the operator instead of subctl without moving chunks of code.

File: pkg/subctl/cmd/createbroker.go [hunk 2]
```diff
@@ -80,5 +84,35 @@ var createBrokerCmd = &cobra.Command{
...
+
+			for _, cond := range found_clusters_crd.Status.Conditions {
+				switch cond.Type {
+				case apiextensionsv1beta1.Established:
+					if cond.Status == apiextensionsv1beta1.ConditionTrue {
+						return true, err
+					}
+				case apiextensionsv1beta1.NamesAccepted:
+					if cond.Status == apiextensionsv1beta1.ConditionFalse {
+						fmt.Printf("Name conflict while wait for CRD Jinghzhu creation: %s, %+v\n", cond.Reason, err)
```
Comment (mangelajo) [inline pkg/subctl/cmd/createbroker.go]: Jinghzhu ? :-)

File: pkg/subctl/cmd/createcrds.go [hunk 3]
```diff
@@ -0,0 +1,41 @@
+package cmd
+
+import (
```
Comment (mangelajo) [inline pkg/subctl/cmd/createcrds.go]: I'd move all this to pkg/broker/creation or something like that. Along with the wait logic.

File: pkg/subctl/cmd/createbroker.go (hunk 1 above)
Comment (dfarrell07) [reply pkg/subctl/cmd/createbroker.go]: Very good idea, I was thinking about early things along these lines, that I wanted to next find a way to structure the logic to make it more logically distinct. I'll take a stab at this pkg approach, thanks!

**Comment by mangelajo** [comment]
I think you guys are duplicating work: https://github.com/submariner-io/submariner-operator/pull/14


I like how @skitt builds the structures, although the comments here also apply there.

**Comment by dfarrell07** [comment]
> I think you guys are duplicating work: #14
> 
> I like how @skitt builds the structures, although the comments here also apply there.

As we clarified on IRC, @skitt's #14 was supposed to be based on his cherry-pick of my subctl patch (#13), but I kept working on that here and broke the base unknowingly. Now I think it's better to just merge #14 and send smaller PRs for future steps. I'll close this to help clarify.

**Comment by skitt** [comment]
It bothers me somewhat that your work on this gets left out, @dfarrell07, but I thought that since I had it working in my version, it was simpler to push that rather than review this, applying the things I’d learned in parallel. I’ll add some notes in review here of non-obvious aspects you might find interesting.

File: pkg/subctl/cmd/createbroker.go [hunk 4]
```diff
@@ -80,5 +84,31 @@ var createBrokerCmd = &cobra.Command{
...
+
+		// Create CRDs
+		fmt.Printf("Creating CRDs\n")
+		apiextensionsClient, err := apiextensionsclientset.NewForConfig(config)
+		//_, err = apiextensionsClient.ApiextensionsV1beta1().CustomResourceDefinitions().Create(createClustersCRD())
+		created_clusters_crd, err := apiextensionsClient.ApiextensionsV1beta1().CustomResourceDefinitions().Create(createClustersCRD())
+		fmt.Printf("Created CRD:\n", created_clusters_crd)
+
+		// Wait for CRD creation
+		err = wait.Poll(5*time.Second, 60*time.Second, func() (bool, error) {
```
Comment (skitt) [inline pkg/subctl/cmd/createbroker.go]: This is missing from #14, so it would be useful to resurrect in another PR. I’m not sure how *necessary* it is in `subctl create broker` itself, since we don’t use the CRDs there, but perhaps we should have a separate `subctl` command to wait for broker readiness (as a general concept — the user shouldn’t know what’s involved).

File: pkg/subctl/cmd/createcrds.go [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+	// Define endpoints.submariner.io CRD spec
+	endpoints_crd_spec_names := apiextensions.CustomResourceDefinitionNames{Plural: "endpoints", Singular: "endpoint", ListKind: "EndpointList", Kind: "Endpoint"}
+	endpoints_crd_spec_versions := apiextensions.CustomResourceDefinitionVersion{Name: "v1", Served: true, Storage: true}
+	endpoints_crd_spec_conversion := apiextensions.CustomResourceConversion{Strategy: "None"}
+	endpoints_crd_spec := apiextensions.CustomResourceDefinitionSpec{Group: "submariner.io", Names: endpoints_crd_spec_names, Scope: "Namespaced", Versions: []apiextensions.CustomResourceDefinitionVersion{endpoints_crd_spec_versions}, Version: "v1", Conversion: &endpoints_crd_spec_conversion}
+
+	// Define endpoints.submariner.io CRD status
+	endpoints_crd_status_names := apiextensions.CustomResourceDefinitionNames{Plural: "endpoints", Singular: "endpoint", ListKind: "EndpointList", Kind: "Endpoint"}
+	endpoints_crd_status_storedversions := []string{"v1"}
+	endpoints_crd_status:= apiextensions.CustomResourceDefinitionStatus{AcceptedNames: endpoints_crd_status_names, StoredVersions: endpoints_crd_status_storedversions}
```
Comment (skitt) [inline pkg/subctl/cmd/createcrds.go]: As I understand it, status shouldn’t be provided on CRD creation, it’s filled in by K8s as the CRD status changes.

File: pkg/subctl/cmd/createcrds.go [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+
+import (
+	apiextensions "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1"
+)
+
+func createEndpointsCRD() *apiextensions.CustomResourceDefinition {
+	// Define endpoints.submariner.io CRD spec
+	endpoints_crd_spec_names := apiextensions.CustomResourceDefinitionNames{Plural: "endpoints", Singular: "endpoint", ListKind: "EndpointList", Kind: "Endpoint"}
+	endpoints_crd_spec_versions := apiextensions.CustomResourceDefinitionVersion{Name: "v1", Served: true, Storage: true}
+	endpoints_crd_spec_conversion := apiextensions.CustomResourceConversion{Strategy: "None"}
```
Comment (skitt) [inline pkg/subctl/cmd/createcrds.go]: These are apparently not necessary.

File: pkg/subctl/cmd/createcrds.go [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+	endpoints_crd_spec := apiextensions.CustomResourceDefinitionSpec{Group: "submariner.io", Names: endpoints_crd_spec_names, Scope: "Namespaced", Versions: []apiextensions.CustomResourceDefinitionVersion{endpoints_crd_spec_versions}, Version: "v1", Conversion: &endpoints_crd_spec_conversion}
+
+	// Define endpoints.submariner.io CRD status
+	endpoints_crd_status_names := apiextensions.CustomResourceDefinitionNames{Plural: "endpoints", Singular: "endpoint", ListKind: "EndpointList", Kind: "Endpoint"}
+	endpoints_crd_status_storedversions := []string{"v1"}
+	endpoints_crd_status:= apiextensions.CustomResourceDefinitionStatus{AcceptedNames: endpoints_crd_status_names, StoredVersions: endpoints_crd_status_storedversions}
+
+	// Define endpoints.submariner.io CRD
+	endpoints_crd := apiextensions.CustomResourceDefinition{Spec: endpoints_crd_spec, Status: endpoints_crd_status}
```
Comment (skitt) [inline pkg/subctl/cmd/createcrds.go]: When trying to use this, it turns out the name has to be given in the metadata as well.
